pkg/network/netlink/consumer.go

```diff
@@ -0,0 +1,410 @@
+package netlink
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+
+	"github.com/pkg/errors"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"golang.org/x/sys/unix"
+
+	"github.com/mdlayher/netlink"
+	"github.com/mdlayher/netlink/nlenc"
+)
+
+const (
+	// netlinkCtNew represents the Netlink multicast group associated to the Conntrack family
+	// representing new connection events. For more information see section "Address formats" in
+	// http://man7.org/linux/man-pages/man7/netlink.7.html
+	netlinkCtNew = uint32(1)
+
+	// ipctnlMsgCtGet represents the Conntrack message type used during the initial load.
+	// This value is defined in include/uapi/linux/netfilter/nfnetlink_conntrack.h
+	ipctnlMsgCtGet = 1
+
+	// outputBuffer is he size of the Consumer output channel.
+	outputBuffer = 100
+
+	// overShootFactor is used sampling rate calculation after the circuit breaker trips.
+	overshootFactor = 0.95
+
+	// netlinkBufferSize is size (in bytes) of the Netlink socket receive buffer
+	// We set it to a large enough size to support bursts of Conntrack events.
+	netlinkBufferSize = 1024 * 1024
+)
+
+var errShortErrorMessage = errors.New("not enough data for netlink error code")
+
+// Consumer is responsible for encapsulating all the logic of hooking into Conntrack via a Netlink socket
+// and streaming new connection events.
+type Consumer struct {
+	conn      *netlink.Conn
+	socket    *Socket
+	pool      *sync.Pool
+	workQueue chan func()
+
+	// rateLimit represents the maximum number of netlink messages per second
+	// that can be read off the netlink socket. Setting it to -1 disables the limit.
+	rateLimit int
```

Comment:
if you're making changes, can this be called `targetRateLimit`

Reply:
Done.